Data race in model.(*folder).startWatch()

```
jb@kvar:~/d/g/s/syncthing $ ./build.sh test
...
==================
WARNING: DATA RACE
Write at 0x00c0004bc720 by goroutine 485:
  github.com/syncthing/syncthing/lib/model.(*folder).startWatch()
      /Users/jb/dev/github.com/syncthing/syncthing/lib/model/folder.go:604 +0x23e
  github.com/syncthing/syncthing/lib/model.(*folder).restartWatch()
      /Users/jb/dev/github.com/syncthing/syncthing/lib/model/folder.go:593 +0x46
  github.com/syncthing/syncthing/lib/model.(*folder).Serve()
      /Users/jb/dev/github.com/syncthing/syncthing/lib/model/folder.go:185 +0x816
  github.com/syncthing/syncthing/lib/model.(*sendOnlyFolder).Serve()
      <autogenerated>:1 +0x43
  github.com/thejerf/suture.(*Supervisor).runService.func1()
      /Users/jb/go/pkg/mod/github.com/thejerf/suture@v0.0.0-20180907184608-bf6ee6a0b047/supervisor.go:540 +0x6c

Previous read at 0x00c0004bc720 by goroutine 555:
  github.com/syncthing/syncthing/lib/model.(*folder).startWatchAsync()
      /Users/jb/dev/github.com/syncthing/syncthing/lib/model/folder.go:642 +0xa16

Goroutine 485 (running) created at:
  github.com/thejerf/suture.(*Supervisor).runService()
      /Users/jb/go/pkg/mod/github.com/thejerf/suture@v0.0.0-20180907184608-bf6ee6a0b047/supervisor.go:528 +0x68
  github.com/thejerf/suture.(*Supervisor).Serve()
      /Users/jb/go/pkg/mod/github.com/thejerf/suture@v0.0.0-20180907184608-bf6ee6a0b047/supervisor.go:426 +0xea2

Goroutine 555 (finished) created at:
  github.com/syncthing/syncthing/lib/model.(*folder).startWatch()
      /Users/jb/dev/github.com/syncthing/syncthing/lib/model/folder.go:606 +0x2cc
  github.com/syncthing/syncthing/lib/model.(*folder).Serve()
      /Users/jb/dev/github.com/syncthing/syncthing/lib/model/folder.go:121 +0x109d
  github.com/syncthing/syncthing/lib/model.(*sendOnlyFolder).Serve()
      <autogenerated>:1 +0x43
  github.com/thejerf/suture.(*Supervisor).runService.func1()
      /Users/jb/go/pkg/mod/github.com/thejerf/suture@v0.0.0-20180907184608-bf6ee6a0b047/supervisor.go:540 +0x6c
==================
--- FAIL: TestPullInvalidIgnoredSO (0.03s)
    testing.go:771: race detected during execution of test
FAIL
```

--- a/lib/model/folder.go
+++ b/lib/model/folder.go
@@ -58,7 +58,7 @@ type folder struct {
 	watchChan        chan []string
 	restartWatchChan chan struct{}
 	watchErr         error
-	watchErrMut      sync.Mutex
+	watchMut         sync.Mutex
 
 	puller puller
 }
@@ -96,7 +96,7 @@ func newFolder(model *Model, cfg config.FolderConfiguration) folder {
 
 		watchCancel:      func() {},
 		restartWatchChan: make(chan struct{}, 1),
-		watchErrMut:      sync.NewMutex(),
+		watchMut:         sync.NewMutex(),
 	}
 }
 
@@ -550,18 +550,18 @@ func (f *folder) scanTimerFired() {
 }
 
 func (f *folder) WatchError() error {
-	f.watchErrMut.Lock()
-	defer f.watchErrMut.Unlock()
+	f.watchMut.Lock()
+	defer f.watchMut.Unlock()
 	return f.watchErr
 }
 
 // stopWatch immediately aborts watching and may be called asynchronously
 func (f *folder) stopWatch() {
+	f.watchMut.Lock()
 	f.watchCancel()
-	f.watchErrMut.Lock()
 	prevErr := f.watchErr
 	f.watchErr = errWatchNotStarted
-	f.watchErrMut.Unlock()
+	f.watchMut.Unlock()
 	if prevErr != errWatchNotStarted {
 		data := map[string]interface{}{
 			"folder": f.ID,
@@ -601,8 +601,10 @@ func (f *folder) startWatch() {
 	f.model.fmut.RLock()
 	ignores := f.model.folderIgnores[f.folderID]
 	f.model.fmut.RUnlock()
+	f.watchMut.Lock()
 	f.watchChan = make(chan []string)
 	f.watchCancel = cancel
+	f.watchMut.Unlock()
 	go f.startWatchAsync(ctx, ignores)
 }
 
@@ -614,10 +616,10 @@ func (f *folder) startWatchAsync(ctx context.Context, ignores *ignore.Matcher) {
 		select {
 		case <-timer.C:
 			eventChan, err := f.Filesystem().Watch(".", ignores, ctx, f.IgnorePerms)
-			f.watchErrMut.Lock()
+			f.watchMut.Lock()
 			prevErr := f.watchErr
 			f.watchErr = err
-			f.watchErrMut.Unlock()
+			f.watchMut.Unlock()
 			if err != prevErr {
 				data := map[string]interface{}{
 					"folder": f.ID,
@@ -639,6 +641,8 @@ func (f *folder) startWatchAsync(ctx context.Context, ignores *ignore.Matcher) {
 				timer.Reset(time.Minute)
 				continue
 			}
+			f.watchMut.Lock()
+			defer f.watchMut.Unlock()
 			watchaggregator.Aggregate(eventChan, f.watchChan, f.FolderConfiguration, f.model.cfg, ctx)
 			l.Debugln("Started filesystem watcher for folder", f.Description())
 			return

